### PR TITLE
Ranking matching models SMN DAM

### PR DESCRIPTION
**Pull request by golovenkov**

### CODE REVIEW

File: deeppavlov/models/ranking/matching_models/dam_utils/operations.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+#   author={Xiangyang Zhou, Lu Li, Daxiang Dong, Yi Liu, Ying Chen, Wayne Xin Zhao, Dianhai Yu and Hua Wu},
+#   booktitle={Proceedings of the 56th Annual Meeting of the Association for Computational Linguistics (Volume 1: Long Papers)},
+#   volume={1},
+#   pages={  --  },
+#   year={2018}
+# }
+# ```
+# http://aclweb.org/anthology/P18-1103
+#
+# Based on authors' Tensorflow code: https://github.com/baidu/Dialogue/tree/master/DAM
```
Comment (yoptar) [inline deeppavlov/models/ranking/matching_models/dam_utils/operations.py]: What about licencing? Is it compatible with Apache 2.0 licence? Can we use this code?

File: deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
+import numpy as np
```
Comment (yoptar) [inline deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py]: Add licence please

File: deeppavlov/models/ranking/matching_models/dam_utils/operations.py (hunk 1 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/matching_models/dam_utils/operations.py]: I was not 100% sure, so I requested the information from authors. It will be clear very soon.

File: deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+@register('ubuntu_dstc7_mt_reader')
+class UbuntuDSTC7MTReader(DatasetReader):
+    """
+    DatasetReader for Ubuntu Dialogue Corpus Dataset (version 3), prepared for DSTC 7 competition Track 1 Subtrack 1.
+
+    https://github.com/IBM/dstc7-noesis
+
+    Args:
+        data_path (str): A path to a folder with dataset csv files.
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py]: yes, json, misspelling

File: deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        where
+        * [context_utt_1, ..., context_utt_10, response_utt_1, ..., response_utt_N] - list that consists of
+        ``num_context_turn`` utterances, followed by ``num_responses`` responses.
+        Where
+        * label - label of the sample
+
+        :param filename: filename to read
+        :param mode: which dataset to return. Can be "train", "valid" or "test"
+        :return: list of contexts and responses with their labels. More details about the format are provided above
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py]: ok 603dbb7

File: deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        """
+        data = []
+        with open(filename, 'rb') as f:
+            json_data = ijson.items(f, 'item')
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py]: you're right 603dbb7

File: deeppavlov/dataset_readers/ubuntu_v2_mt_reader.py [hunk 6]
```diff
@@ -70,13 +73,33 @@ def preprocess_data_validation(self, fname: Union[Path, str]) -> List[Tuple[List
             reader = csv.reader(f)
             next(reader)
             for el in reader:
-                contexts.append(self._expand_context(el[0].split('__eot__')))
+                contexts.append(self._expand_context(el[0].split('__eot__'), padding=self.padding))
                 responses.append(el[1:])
         data = [el[0] + el[1] for el in zip(contexts, responses)]
-        data = [(el, 1) for el in data]
+        data = [(el, 1) for index, el in enumerate(data)]   # NOTE: labels are useless here actually...
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply deeppavlov/dataset_readers/ubuntu_v2_mt_reader.py]: i used to, but not now, thanks 603dbb7

File: deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            filename (Path): filename to read
+            mode (str): which dataset to return. Can be "train", "valid" or "test"
+
+        Returns:
+             list of contexts and responses with their labels. More details about the format are provided above
+        """
+        data = []
+        with open(filename, 'rb') as f:
+            json_data = json.load(open(filename, 'rt'))
+            for entry in json_data:
```
Comment (yoptar) [inline deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py]:
```python
with open(filename, encoding='utf8') as f:
    json_data = json.load(f)

for entry in json_data:
    ...
```
When `open` not as bytes, encoding is necessary because on different systems it defaults to different values.

File: deeppavlov/models/preprocessors/assemble_embeddings_matrix.py [hunk 8]
```diff
@@ -88,7 +88,7 @@ def __init__(self,
                         self.emb_mat[n] = pca(embedder([[token]])[0])[0]
                     else:
                         self.emb_mat[n] = embedder([[token]])[0][0]
-                except KeyError:
+                except (KeyError, IndexError):
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply deeppavlov/models/preprocessors/assemble_embeddings_matrix.py]: should not obviously, thanks 603dbb7

File: deeppavlov/models/preprocessors/ubuntu_v2_preprocessor.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            If set to ``post`` it will truncated at the end.
+        num_context_turns: A number of ``context`` turns in data samples.
+        num_ranking_samples: A number of condidates for ranking including positive one.
+        tokenizer: An instance of one of the :class:`deeppavlov.models.tokenizers`.
+        vocab: An instance of :class:`deeppavlov.core.data.simple_vocab.SimpleVocabulary`.
+    """
+
+    def __init__(self,
+                 save_path: str = './tok.dict',
+                 load_path: str = './tok.dict',
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply deeppavlov/models/preprocessors/ubuntu_v2_preprocessor.py]: not sure if I can remove the variables, so I just removed its default values. There are full paths for save_path and load_path in related configs 603dbb7

File: deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py (hunk 2 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py]: I was not 100% sure, so I requested the information from authors. It will be clear very soon, in a few days. It seems that we should to wait some time until the approval.

File: deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+from deeppavlov.core.common.log import get_logger
+from deeppavlov.core.common.registry import register
+from deeppavlov.models.ranking.matching_models .tf_base_matching_model import TensorflowBaseMatchingModel
+
+import tensorflow as tf
```
Comment (yoptar) [inline deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py]: imports order, pep 8

File: deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+from deeppavlov.core.common.log import get_logger
+from deeppavlov.core.common.registry import register
+from deeppavlov.models.ranking.matching_models .tf_base_matching_model import TensorflowBaseMatchingModel
```
Comment (yoptar) [inline deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py]: excess space

File: deeppavlov/models/ranking/matching_models/sequential_matching_network.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+                 max_num_utterance: int = 10,
+                 max_sequence_length: int = 50,
+                 learning_rate: float = 1e-3,
+                 emb_matrix: np.ndarray = None,
+                 trainable_embeddings: bool = False,
+                 *args,
+                 **kwargs):
+
+
+        self.seed = seed
```
Comment (yoptar) [inline deeppavlov/models/ranking/matching_models/sequential_matching_network.py]: Is `self.seed` used?

Comment (golovenkov) [reply deeppavlov/models/ranking/matching_models/sequential_matching_network.py]: actually, no

File: deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py (hunk 11 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py]: done 951e205

Comment (yoptar) [reply deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py]: `matching_models .tf_base_matching_model`

**Comment by yoptar** [comment]
New configs should be added to tests

File: deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py (hunk 11 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/matching_models/deep_attention_matching_network.py]: copypasted (
done 3aafafc

File: docs/intro/features.rst [hunk 13]
```diff
@@ -184,23 +184,37 @@ on Automatic Spelling Correction for Russian:
...
+   +-------------------+----------------------------------------------------------------------------------------------------------------+-----------------------+------------------+
+   |    Dataset        | Model config                                                                                                   | Validation (Recall@1) | Test1 (Recall@1) |
+   +===================+================================================================================================================+=======================+==================+
+   | `InsuranceQA V1`_ | :config:`ranking_insurance_interact <ranking/ranking_insurance_interact.json>`                                 |   72.0                |   72.2           |
+   +-------------------+----------------------------------------------------------------------------------------------------------------+-----------------------+------------------+
+   | `Ubuntu V2`_      | :config:`ranking_ubuntu_v2_mt_word2vec_dam_interact <ranking/ranking_ubuntu_v2_mt_word2vec_dam_interact.json>` |   70.3                |   69.8           |
```
Comment (golovenkov) [inline docs/intro/features.rst]: Should we add r@2, r@5 metric values for all ranking models as well?

File: deeppavlov/models/ranking/matching_models/tf_base_matching_model.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+from typing import List, Iterable, Union
+
+from deeppavlov.core.common.log import get_logger
+from deeppavlov.models.ranking.siamese_model import SiameseModel
+from deeppavlov.core.models.tf_model import TFModel
+
+log = get_logger(__name__)
+
+
+class TensorflowBaseMatchingModel(TFModel, SiameseModel):
```
Comment (yoptar) [inline deeppavlov/models/ranking/matching_models/tf_base_matching_model.py]: If this is an abstract class, it should be marked as such
```suggestion
class TensorflowBaseMatchingModel(TFModel, SiameseModel, metaclass=ABC):
```

File: deeppavlov/models/ranking/matching_models/tf_base_matching_model.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+    Args:
+        none
+
+    """
+
+    def __init__(self,
+                 *args,
+                 **kwargs):
+        super(TensorflowBaseMatchingModel, self).__init__(*args, **kwargs)
```
Comment (yoptar) [inline deeppavlov/models/ranking/matching_models/tf_base_matching_model.py]: `__init__` that does nothing can be skipped

File: deeppavlov/models/ranking/matching_models/dam_utils/operations.py (hunk 1 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/matching_models/dam_utils/operations.py]: 1) Authors of DAM network would like to use MIT license, they think it should be compatible with Apache 2.0 license.
2) SMN license is not approved yet

File: deeppavlov/models/ranking/matching_models/dam_utils/operations.py [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+
+def learning_rate(step_num, d_model=512, warmup_steps=4000):
+    a = step_num**(-0.5)
+    b = step_num*warmup_steps**(-1.5)
+    return a, b, d_model**(-0.5) * min(step_num**(-0.5), step_num*(warmup_steps**(-1.5))) 
+
+def selu(x):
+    alpha = 1.6732632423543772848170429916717
+    scale = 1.0507009873554804934193349852946
+    print('use selu')
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply deeppavlov/models/ranking/matching_models/dam_utils/operations.py]: done 32fa6ec

File: docs/intro/features.rst (hunk 13 above)
Comment (puleon) [reply docs/intro/features.rst]: I think it's worth doing. You could now fill cells for your models and leave them empty for my models.

File: deeppavlov/models/preprocessors/ubuntu_v2_preprocessor.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.models.component import Component
+from deeppavlov.core.models.estimator import Estimator
+from deeppavlov.core.data.utils import zero_pad_truncate
+
+log = get_logger(__name__)
+
+
+@register('ubuntu_v2_preprocessor')
+class UbuntuV2Preprocessor(Estimator):
```
Comment (puleon) [inline deeppavlov/models/preprocessors/ubuntu_v2_preprocessor.py]: Why is necessity for this class? It looks like the duplicate of the SiamesePreprocessor class.

File: docs/components/neural_ranking.rst [hunk 18]
```diff
@@ -34,6 +34,34 @@ command:
 As an example of configuration file see
 :config:`ranking_insurance.json <ranking/ranking_insurance.json>`.
 
+
+To use ether Sequential Matching Network (SMN) or Deep Attention Matching Network (DAM)
```
Comment (puleon) [inline docs/components/neural_ranking.rst]:
```suggestion
To use either Sequential Matching Network (SMN) or Deep Attention Matching Network (DAM)
```

File: deeppavlov/models/ranking/matching_models/tf_base_matching_model.py [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+import numpy as np
+from typing import List, Iterable, Union
+
+from deeppavlov.core.common.log import get_logger
+from deeppavlov.core.models.tf_model import TFModel
+
+log = get_logger(__name__)
+
+
+class TensorflowBaseMatchingModel(TFModel):
```
Comment (puleon) [inline deeppavlov/models/ranking/matching_models/tf_base_matching_model.py]: The functional of this class repeats the `SiameseModel` functional. Instead of rewriting the __call__ complicated method one can easily override the `_make_batch_function`.

File: deeppavlov/models/ranking/matching_models/tf_base_matching_model.py [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+                batch_buffer_context_len += lens
+
+                lens = []
+                for context in [response_sentence for response_sentence in response_sentences]:
+                    lens.append(len(context[context != 0]))
+                batch_buffer_response_len += lens
+
+                if len(batch_buffer_context) >= self.batch_size:
+                    for i in range(len(batch_buffer_context) // self.batch_size):
+                        feed_dict = {
```
Comment (puleon) [inline deeppavlov/models/ranking/matching_models/tf_base_matching_model.py]: This is the duplication of the code of `__call__` method.

Comment (puleon) [inline deeppavlov/models/ranking/matching_models/tf_base_matching_model.py]: https://github.com/deepmipt/DeepPavlov/blob/14266e0627eb8675ea51da21697a007d39a9dfba/deeppavlov/models/ranking/matching_models/tf_base_matching_model.py#L165-L175 This is duplication of the `__call__` method code.

File: deeppavlov/models/ranking/matching_models/matching_predictor.py [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+from deeppavlov.core.common.log import get_logger
+from deeppavlov.core.models.component import Component
+from deeppavlov.models.ranking.matching_models.tf_base_matching_model import TensorflowBaseMatchingModel
+from deeppavlov.core.common.registry import register
+
+log = get_logger(__name__)
+
+
+@register('matching_predictor')
+class MatchingPredictor(Component):
```
Comment (puleon) [inline deeppavlov/models/ranking/matching_models/matching_predictor.py]: Why is necessity for this class? It is very similar to `SiamesePredictor` class.

**Comment by seliverstov** [comment]
Add a short description and list of the key features to the head of the pull request.

**Comment by seliverstov** [comment]
Closed due to author inactivity.
